File: 199/main.py
from typing import Optional, List
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        """
        approach
        for every level, put onto new queue
        result.append(queue[-1])
        queue = new_queue
        end when queue empty
        """
        result = []
        if root is None:
            return result
        queue = deque([root])
        while queue:
            result.append(queue[-1].val)
            new_queue = deque()
            for node in queue:
                if node.left is not None:
                    new_queue.append(node.left)
                if node.right is not None:
                    new_queue.append(node.right)
            queue = new_queue
        return result

File: 199/main.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function rightSideView(root: TreeNode | null): number[] {
  if (root === null) return [];
  let results: number[] = [];
  let currentRow: TreeNode[] = [root];
  while (currentRow.length !== 0) {
    results.push(currentRow[currentRow.length - 1].val);
    let newList: TreeNode[] = [];
    for (let node of currentRow) {
      if (node.left) newList.push(node.left);
      if (node.right) newList.push(node.right);
    }
    currentRow = newList;
  }
  return results;
}

// approach
// list of results
// list of root
// while list is not empty
// results.append(list[list.length - 1])
// let newList = []
// for every node in list
// newList.push(node.left, node.right)
// list = newList
